### DISCUSSION TITLE
Lazy Fibonacci question

### DISCUSSION THREAD

**Discussion by thedeemon** [flix/flix#6250] [General]
Hi, I was looking at code samples on the main page of flix.dev and saw the Fibonacci example.

In Haskell with its lazy lists, when we define it as
```
fibs = 0 : 1 : zipWith (+) fibs (tail fibs)
```
the `fibs` on the right refers to the same list `fibs` we define on the left. It's a value, not a function. So if I ask one million-th element of that list, the list will unroll in memory once to 1M elements.

However in the Flix example
```
def fibs(): DelayList[Int32] =
    DelayList.LCons(0,
        lazy DelayList.LCons(1,
            lazy DelayList.zipWith(
                (x, y) -> x + y, fibs(), DelayList.tail(fibs()))))
```
it's a function and to me it looks like every call to it returns a new list instead of reusing and unrolling one lazy list. Am I missing something about Flix functions (maybe there is some optimisation for pure functions I'm not aware of) or this particular example? What is the actual O(...) complexity of this function?

**Comment by magnus-madsen** [comment]
Yes, you are right. In the example `fibs` is a function, so if you call it twice, it will build two lists.

We could use the fact that its pure and that it is a `Unit` function to memoize it, but we do not.

**Reply by thedeemon** [reply]
Thanks! It calls itself twice recursively, so it's probably exponential. I wonder how this sample shall be changed to make it linear as in Haskell. In OCaml there is `let rec` in addition to `let`, and in Haskell ordinary `let` is recursive. Is `let` in Flix recursive?

**Reply by magnus-madsen** [reply]
A local def is letrec, but I am not sure if it would work here.
